modules/light-clients/07-tendermint/consensus_host_test.go

```diff
@@ -0,0 +1,239 @@
+package tendermint_test
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	commitmenttypes "github.com/cosmos/ibc-go/v8/modules/core/23-commitment/types"
+	"github.com/cosmos/ibc-go/v8/modules/core/exported"
+	solomachine "github.com/cosmos/ibc-go/v8/modules/light-clients/06-solomachine"
+	ibctm "github.com/cosmos/ibc-go/v8/modules/light-clients/07-tendermint"
+	ibctesting "github.com/cosmos/ibc-go/v8/testing"
+	"github.com/cosmos/ibc-go/v8/testing/mock"
+)
+
+func (suite *TendermintTestSuite) TestGetSelfConsensusState() {
+	var height clienttypes.Height
+
+	cases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			name: "zero height",
+			malleate: func() {
+				height = clienttypes.ZeroHeight()
+			},
+			expError: clienttypes.ErrInvalidHeight,
+		},
+		{
+			name: "height > latest height",
+			malleate: func() {
+				height = clienttypes.NewHeight(1, uint64(suite.chainA.GetContext().BlockHeight())+1)
+			},
+			expError: stakingtypes.ErrNoHistoricalInfo,
+		},
+		{
+			name: "custom consensus host: failure",
+			malleate: func() {
+				consensusHost := &mock.ConsensusHost{
+					GetSelfConsensusStateFn: func(ctx sdk.Context, height exported.Height) (exported.ConsensusState, error) {
+						return nil, mock.MockApplicationCallbackError
+					},
+				}
+				suite.chainA.GetSimApp().GetIBCKeeper().ClientKeeper.SetSelfConsensusHost(consensusHost)
+			},
+			expError: mock.MockApplicationCallbackError,
+		},
+		{
+			name: "custom consensus host: success",
+			malleate: func() {
+				consensusHost := &mock.ConsensusHost{
+					GetSelfConsensusStateFn: func(ctx sdk.Context, height exported.Height) (exported.ConsensusState, error) {
+						return &solomachine.ConsensusState{}, nil
+					},
+				}
+				suite.chainA.GetSimApp().GetIBCKeeper().ClientKeeper.SetSelfConsensusHost(consensusHost)
+			},
+			expError: nil,
+		},
+		{
+			name: "latest height - 1",
+			malleate: func() {
+				height = clienttypes.NewHeight(1, uint64(suite.chainA.GetContext().BlockHeight())-1)
+			},
+			expError: nil,
+		},
+		{
+			name: "latest height",
+			malleate: func() {
+				// historical info is set on BeginBlock in x/staking, which is now encapsulated within the FinalizeBlock abci method,
+				// thus, we do not have historical info for current height due to how the ibctesting library operates.
+				// ibctesting calls app.Commit() as a final step on NextBlock and we invoke test code before FinalizeBlock is called at the current height once again.
```

Comment:
Found some unexpected behaviour here with how testing lib operates. 

Invoking `TrackHistoricalInfo` directly resolves this for the time being.

Reply:
Opened https://github.com/cosmos/ibc-go/issues/6024